tests/subsys/power/power_mgmt/interface_prj.conf

```diff
@@ -0,0 +1,11 @@
+CONFIG_ZTEST=y
+CONFIG_PM=y
+CONFIG_PM_DEVICE=y
+CONFIG_PM_DEVICE_IDLE=y
+CONFIG_PM_POLICY_APP=y
+CONFIG_MP_NUM_CPUS=1
+CONFIG_PM_DEBUG=y
+CONFIG_NO_OPTIMIZATIONS=y
```

Comment:
This is the same hack in a different place though.

Can you detail exactly why you need this?  It sounds like you've discovered a bug in the coverage code generation somewhere, and all this is doing is papering over the problem.  Let's not do this.  If we can't commit this cleanly because there's a bug in coverage let's get the bug in coverage filed and fixed first.

Reply:
In subsys/power/device.c there is a weak variable z_pm_core_devices, I define a strong version in my test file. on platform qemu_x86_64 run my test case with --coverage everything is fine, but fail when run them without --coverage, by examining generated zephyr.lst, for lines 153-159 in function "void pm_create_device_list(void)":
153                 while (z_pm_core_devices[cdi]) {                                                                                                                                                                 154                         if (strcmp(dev->name, z_pm_core_devices[cdi]) == 0) {                                                                                                                                    155                                 pm_devices[cdi] = pmi;                                                                                                                                                           156                                 break;                                                                                                                                                                           157                         }                                                                                                                                                                                        158                         ++cdi;                                                                                                                                                                                   159                 } 
in zephyr.lst with --coverage:
**24703         while (z_pm_core_devices[core_dev]) {
24704   1093b8:       eb 1d                   jmp    1093d7 <pm_create_device_list+0xcc>
24705                 core_dev++;
24706   1093ba:       0f b7 45 fc             movzwl -0x4(%rbp),%eax
24707   1093be:       83 c0 01                add    $0x1,%eax
24708   1093c1:       66 89 45 fc             mov    %ax,-0x4(%rbp)
24709   1093c5:       48 8b 05 9c ca 02 00    mov    0x2ca9c(%rip),%rax        # 135e68 <__gcov0.pm_create_device_list+0x28>
24710   1093cc:       48 83 c0 01             add    $0x1,%rax
24711   1093d0:       48 89 05 91 ca 02 00    mov    %rax,0x2ca91(%rip)        # 135e68 <__gcov0.pm_create_device_list+0x28>
24712         while (z_pm_core_devices[core_dev]) {
24713   1093d7:       0f b7 45 fc             movzwl -0x4(%rbp),%eax
24714   1093db:       48 98                   cltq
24715   1093dd:       48 8b 04 c5 70 fb 12    mov    0x12fb70(,%rax,8),%rax
24716   1093e4:       00
24717   1093e5:       48 85 c0                test   %rax,%rax
24718   1093e8:       75 d0                   jne    1093ba <pm_create_device_list+0xaf>
24719         }**
while in zephyr.lst without --coverage:
**16431         while (z_pm_core_devices[core_dev]) {
16432   101ba5:       0f b7 c3                movzwl %bx,%eax
16433   101ba8:       48 83 3c c5 30 b3 10    cmpq   $0x0,0x10b330(,%rax,8)
16434   101baf:       00 00
16435   101bb1:       74 04                   je     101bb7 <pm_create_device_list+0x5a>
16436                 core_dev++;
16437   101bb3:       ff c3                   inc    %ebx
16438   101bb5:       eb ee                   jmp    101ba5 <pm_create_device_list+0x48>
16439         }**
there must be something be optimized out and cause test cases failure. "--coverage" avoid this optimization by implicitly disable "CONFIG_NO_OPTIMIZATIONS" and modify some configurations such as "CONFIG_IDLE_STACK_SIZE", 
"CONFIG_TEST_EXTRA_STACKSIZE", "CONFIG_SYSTEM_WORKQUEUE_STACK_SIZE".
I have tried add "__attribute__((optimize("O0")))" before "void pm_create_device_list(void)", it works, but ugly.